File: maven/org.friendularity.spec.connection/src/main/java/org/friendularity/spec/connection/ServiceBindingSpec.scala
 /*
 * Copyright 2013 The Friendularity Project (www.friendularity.org).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.friendularity.spec.connection

import org.appdapter.core.component.KnownComponentImpl
import org.appdapter.bind.rdf.jena.assembly.CachingComponentAssembler
import org.appdapter.core.item.{Item}
import org.appdapter.core.name.{Ident}
import com.hp.hpl.jena.assembler.{Assembler, Mode}
import scala.collection.mutable

/**
 * The data object for the dependency binding properties of a service.
 *
 * @author Jason Randolph Eads <eadsjr@hansonrobokind.com>
 */
class ServiceBindingSpec extends KnownComponentImpl {

  // The stored data definitions
  var className: String
  val properties = Map.empty[String,String]
  
  // Setters for data
  def setClassName(classname: String) { className = classname }
  
  // Accumulators for data
  def addProperty(key:String, value:String) { properties + (key->value) }
  
  // Getters for data
  def getClassName(): String = { className }
  def getProperties(): Map[String,String] = { properties }
}

/**
 * The builder that is called to make a Spec object from the raw RDF data
 * representing a ServiceBinding.
 * 
 * @author Jason Randolph Eads <eadsjr@hansonrobokind.com>
 */
 class ServiceBindingSpecBuilder extends CachingComponentAssembler[ServiceBindingSpec] {

  override protected def decideComponentClass(ident: Ident, item: Item): Class[ServiceBindingSpec] = {
    return classOf[ServiceBindingSpec]
  }
  
  override protected def initExtendedFieldsAndLinks(sbSpec: ServiceBindingSpec, configItem : Item, assmblr : Assembler , mode: Mode ) {
    val itemAssemblyReader = getReader()
    
    // Defines the relationship "#Property Name" key (aka the Predicate), that is followed from an individual to collect the data
    val serviceClassNameId = "serviceJavaFQCN"
    val servicePropertiesId = "hasProperty";
    
    // read in the data field and store it in the Spec
    sbSpec.setClassName(itemAssemblyReader.readConfigValString(configItem.getIdent(), serviceClassNameId, configItem, ""))
    
    // read in and build the linked properties, and storing each in the Spec
    val linkedProperties : java.util.List[Object] = itemAssemblyReader.findOrMakeLinkedObjects(configItem, servicePropertiesId, assmblr, mode, null)
    for( o <- linkedProperties) {
      o match {
        case propertySpec: PropertySpec => sbSpec.addProperty(propertySpec.getName, propertySpec.getValue)
        case _ => getLogger.warn("Unexpected object found at {} = {}", Array[Object]( servicePropertiesId, o));
      }
    }
   }
 }
 

File: maven/org.friendularity.spec.connection/src/main/java/org/friendularity/spec/connection/ServiceBindingSpec.java
package org.friendularity.spec.connection;

import java.util.HashMap;
import java.util.Map;
import org.appdapter.core.component.KnownComponentImpl;
import org.jflux.api.registry.Descriptor;
import org.jflux.api.registry.basic.BasicDescriptor;
import org.jflux.api.service.ServiceDependency;
import org.jflux.api.service.ServiceDependency.Cardinality;
import org.jflux.api.service.ServiceDependency.UpdateStrategy;
import org.jflux.api.service.binding.ServiceBinding.BindingStrategy;

/**
 * The data object for the dependency binding properties of a service.
 *
 * @author Jason Randolph Eads <eadsjr@hansonrobokind.com>
 * @author Jason G. Pallack <jgpallack@gmail.com>
 */

public class ServiceBindingSpec extends KnownComponentImpl {
    // The stored data definitions
    
    private String myClassName;
    private Map<String, String> myProperties;
    private Cardinality myCardinality;
    private UpdateStrategy myUpdateStrategy;
    
    private BindingStrategy myBindingStrategy;
    
    public ServiceBindingSpec() {
        myProperties = new HashMap<String, String>();
        myCardinality = null;
        myUpdateStrategy = null;
    }
    
    // Getters for data
    
    public ServiceDependency getServiceDependency() {
        return new ServiceDependency(
                myClassName, myClassName, myCardinality, myUpdateStrategy,
                myProperties);
    }
    
    public Descriptor getDescriptor() {
        return new BasicDescriptor(myClassName, myProperties);
    }
    
    public BindingStrategy getBindingStrategy() {
        return myBindingStrategy;
    }
    
    // Setters for data
    
    public void setClassName(String className) {
        myClassName = className;
    }
    
    public void setBindingStrategy(BindingStrategy bindingStrategy) {
        myBindingStrategy = bindingStrategy;
    }
    
    public void setCardinality(Cardinality cardinality) {
        myCardinality = cardinality;
    }
    
    public void setUpdateStrategy(UpdateStrategy updateStrategy) {
        myUpdateStrategy = updateStrategy;
    }
    
    // Accumulators for data
    
    public void addProperty(String key, String value) {
        myProperties.put(key, value);
     }
     
    public void removeProperty(String key) {
        myProperties.remove(key);
    }
}
